src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs

```diff
@@ -159,6 +188,316 @@ private string GetRootPathWithGuid_NoLock()
             return new MetadataAsSourceFile(fileInfo.TemporaryFilePath, navigateLocation, documentName, documentTooltip);
         }
 
+        private async Task<Document> DecompileSymbolAsync(Document temporaryDocument, OptionSet options, ISymbol symbol, CancellationToken cancellationToken)
+        {
+            // Get the name of the type the symbol is in
+            var containingOrThis = symbol.GetContainingTypeOrThis();
+            var fullName = GetFullReflectionName(containingOrThis);
+
+            var compilation = await temporaryDocument.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            // TODO: retrieve path to actual assembly instead of reference assembly
+            var reference = compilation.GetMetadataReference(symbol.ContainingAssembly);
+            // Load the assembly.
+            var ad = AssemblyDefinition.ReadAssembly(reference.Display, new ReaderParameters() { AssemblyResolver = new RoslynAssemblyResolver(compilation) });
+
+            var decompiler = new CSharpDecompiler(ad.MainModule, GetDecompilerSettings(options));
+            var fullTypeName = new FullTypeName(fullName);
+
+            // Try to decompile; if an exception is thrown the caller will handle it
+            var text = decompiler.DecompileTypeAsString(fullTypeName);
+            return temporaryDocument.WithText(SourceText.From(text));
+        }
+
+        private static DecompilerSettings GetDecompilerSettings(OptionSet options)
+        {
+            var decompilerSettings = new DecompilerSettings();
+            var formattingOptions = decompilerSettings.CSharpFormattingOptions;
+
+            // Indentation
+            formattingOptions.IndentNamespaceBody = true;
+            formattingOptions.IndentClassBody = true;
+            formattingOptions.IndentInterfaceBody = true;
+            formattingOptions.IndentStructBody = true;
+            formattingOptions.IndentEnumBody = true;
+            formattingOptions.IndentMethodBody = true;
+            formattingOptions.IndentPropertyBody = true;
+            formattingOptions.IndentEventBody = true;
```

Comment:
A lot of these match the defaults that teh library provides: https://github.com/icsharpcode/NRefactory/blob/master/ICSharpCode.NRefactory.CSharp/Formatter/FormattingOptionsFactory.cs

Can we eschew the redundant initializers?

Reply:
I was intentionally verbose so it's clear that we're not skipping any by accident. This form makes it clear which options map to a setting, and which options only map to a hard-coded default (or otherwise do not map to an option).

Reply:
Gotcha.  That seems reasonable to me.  Possibly worth a comment explaining such.